Zend/Optimizer/zend_func_info.c

```diff
@@ -834,6 +834,72 @@ static const func_info_t func_infos[] = {
 static HashTable func_info;
 ZEND_API int zend_func_info_rid = -1;
 
+void zend_verify_internal_func_info(zend_function *fn, zval *retval) {
+	if (fn->common.scope) {
+		/* This is a method, not a function. */
+		return;
+	}
+
+	zend_string *name = fn->common.function_name;
+	if (!name) {
+		/* The pass function has no name. */
+		return;
+	}
+
+	zval *zv = zend_hash_find_ex(&func_info, name, 1);
+	if (!zv) {
+		return;
+	}
+
+	func_info_t *info = Z_PTR_P(zv);
+	if (info->info_func) {
+		return;
+	}
+
+	uint32_t type_mask = info->info;
+	if (Z_REFCOUNTED_P(retval)) {
+		if (!(type_mask & MAY_BE_RC1)) {
+			zend_error_noreturn(E_CORE_ERROR, "%s() missing rc1", ZSTR_VAL(name));
+		}
+		if (Z_REFCOUNT_P(retval) > 1 && !(type_mask & MAY_BE_RCN)) {
+			zend_error_noreturn(E_CORE_ERROR, "%s() missing rcn", ZSTR_VAL(name));
+		}
+	}
+
+	uint32_t type = 1u << Z_TYPE_P(retval);
+	if (!(type_mask & type)) {
+		zend_error_noreturn(E_CORE_ERROR, "%s() missing type", ZSTR_VAL(name));
+	}
+
+	if (Z_TYPE_P(retval) == IS_ARRAY) {
+		/* Don't check large arrays. */
+		HashTable *ht = Z_ARRVAL_P(retval);
+		if (zend_hash_num_elements(ht) < 16) {
```

Comment:
Instead of not checking large arrays, another approach to consider would be to check the first 15 elements of any array for more debug coverage?
Or explain that it's deliberately not done for performance or other reasons

```c
if (num_checked >= 15) {
   break;
}
num_checked++;
```